sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala

```diff
@@ -0,0 +1,471 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution
+
+import java.util
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans.physical.{AllTuples, Distribution, ClusteredDistribution, Partitioning}
+import org.apache.spark.util.collection.CompactBuffer
+
+case class Window(
+    projectList: Seq[Attribute],
+    windowExpression: Seq[NamedExpression],
+    windowSpec: WindowSpecDefinition,
+    child: SparkPlan)
+  extends UnaryNode {
+
+  override def output: Seq[Attribute] =
+    (projectList ++ windowExpression).map(_.toAttribute)
+
+  override def requiredChildDistribution: Seq[Distribution] =
+    if (windowSpec.partitionSpec.isEmpty) {
+      // This operator will be very expensive.
+      AllTuples :: Nil
+    } else {
+      ClusteredDistribution(windowSpec.partitionSpec) :: Nil
+    }
+
+  // Since window functions are adding columns to the input rows, the child's outputPartitioning
+  // is preserved.
+  override def outputPartitioning: Partitioning = child.outputPartitioning
+
+  override def requiredChildOrdering: Seq[Seq[SortOrder]] = {
+    // For now, the required child ordering has two parts.
+    // The first part is the expressions in the partition specification.
+    // The second part is the expressions specified in the ORDER BY cluase.
+    // Basically, we first use sort to group rows based on partition specifications and then sort
+    // Rows in a group based on the order specification.
+    (windowSpec.partitionSpec.map(SortOrder(_, Ascending)) ++ windowSpec.orderSpec) :: Nil
```

Comment:
if partition key is ordered, we can easily get rows with the same partition key in the execution, otherwise we need redistribute them in memory. It is too expensive